### PR TITLE
fix(*): new serialization mechanism

### PR DESCRIPTION
**Pull request by garrappachc**
Using class-transformer to serialize objects does not allow us to use async methods, thus rendering us incapable of resolving sub-documents properly. This PR attemps to approach object serialization another way, using proper services, some meta-types and injector.

Model changes:
- `Game.slot.player` is a `Player` instance,
- `Game.gameServer` is a `GameServer` instance,
- `QueueSlot.playerId` is gone

### CODE REVIEW

**Comment by c43721** [comment]
This is kinda why I hate JS. I can't really just automatically assume serialization fields with async or can I just "return and forget"

While this approach looks nice, you're basically just using classes. Why not just use classes instead of creating an entirely abstract interface + paying for the injection costs?

Something like `return this.toJSON()`  in the classes? And if you need more information than just do something like `return {...this.toJSON(), someField: someService.toJSON()`

**Comment by garrappachc** [comment]
It's not a limitation of JS, it's class-transformer's one. [The issue is reported](https://github.com/typestack/class-transformer/issues/549), but it seems any kind of development is on hold for now, sadly.

When it comes to your second question: yes, we could go for a full object-oriented approach, but it seems like pissing into the wind with NestJS' architecture. We would have to workaround the whole injection system, it looks... nasty?

```ts
export class Game implements Serializable<GameDto> {
  // ...
  async serialize(): Promise<GameDto> {
    const playersService = app.get(PlayersService); // this line here seems wrong
    return {
      // ...
      players: Promise.all(this.players.map(async player => await (await playersService.getById(player)).serialize()),
      // ...
    };
  }
}
```

**Comment by garrappachc** [comment]
I mean, I'm not _entirely_ against this approach, I just think it's a little bit more iffy...

**Comment by c43721** [comment]
everything about serialization is stupid.

**Comment by garrappachc** [comment]
> everything about serialization is stupid.

That's a strong statement if I've ever seen one...

**Comment by garrappachc** [comment]
On the other hand, `app.get()` is a dependency injection as well, so I guess there's nothing wrong with the second approach...

**Comment by garrappachc** [comment]
@c43721 take a look now, does it look better to you?

**Comment by c43721** [comment]
I actually like this newer approach. That's quite smart.

**Comment by c43721** [comment]
Technically we can use intercepters and return instance of like some class (Serializable in our case) and do the `.serialize()` in an intercepter if we don't want to call it ourselves.

**Comment by garrappachc** [comment]
That would require traversing deep through object to find instances of `Serializable`... Efficiency where?

**Comment by c43721** [comment]
> js

> efficiency

lol

(it would be at most like 20ms additional time. You're using express- you don't care about performance. If you did, we'd be using fastify and not nest)

**Comment by garrappachc** [comment]
@c43721 ready for review. The tests are a mess, don't look at them.

File: src/auth/strategies/steam.strategy.spec.ts [hunk 1]
```diff
@@ -5,14 +5,12 @@ import { Test, TestingModule } from '@nestjs/testing';
...
-  hasAcceptedRules: true,
-  roles: [],
-  _links: [],
-};
+const mockPlayer = new Player();
+mockPlayer.id = 'FAKE_PLAYER_ID';
+mockPlayer.name = 'FAKE_PLAYER';
+mockPlayer.steamId = 'FAKE_STEAM_ID';
+mockPlayer.hasAcceptedRules = true;
+mockPlayer.roles = [];
```
Comment (c43721) [inline src/auth/strategies/steam.strategy.spec.ts]: Why not define these properties within the `Player()` constructor instead of modifying this?

File: src/games/controllers/games-with-substitution-requests.controller.ts [hunk 2]
```diff
@@ -1,18 +1,16 @@
...
 import { GamesService } from '../services/games.service';
 
 @Controller('games-with-substitution-requests')
 export class GamesWithSubstitutionRequestsController {
   constructor(private gamesService: GamesService) {}
 
   @Get()
-  @UseInterceptors(ClassSerializerInterceptor)
-  async getGamesWithSubstitutionRequests() {
+  @UseInterceptors(SerializerInterceptor)
```
Comment (c43721) [inline src/games/controllers/games-with-substitution-requests.controller.ts:12]: Instead of doing this on each controller, can we just use global interceptors? Since we ignore anything that isn't instance of `Serializer` I don't think this will hurt. It would clean up our code and enforce this everywhere, making it much safer in the future IMHO.

File: src/games/controllers/games-with-substitution-requests.controller.ts [hunk 3]
```diff
@@ -1,18 +1,16 @@
...
+  async getGamesWithSubstitutionRequests(): Promise<Serializable<GameDto>[]> {
```
Comment (c43721) [inline src/games/controllers/games-with-substitution-requests.controller.ts:13]: Can this not be inferred?

File: src/games/dto/game.dto.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  number: number;
+  slots: {
+    player: Serializable<PlayerDto>;
+    team: 'red' | 'blu';
+    gameClass: Tf2ClassName;
+    status: 'active' | 'waiting for substitute' | 'replaced';
+    connectionStatus: 'offline' | 'joining' | 'connected';
+  }[];
+  map: string;
+  state: 'launching' | 'started' | 'ended' | 'interrupted';
```
Comment (c43721) [inline src/games/dto/game.dto.ts:18]: nit: convert to enum

File: src/games/dto/game.dto.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  id: string;
+  launchedAt: Date;
+  endedAt?: Date;
+  number: number;
+  slots: {
+    player: Serializable<PlayerDto>;
+    team: 'red' | 'blu';
+    gameClass: Tf2ClassName;
+    status: 'active' | 'waiting for substitute' | 'replaced';
+    connectionStatus: 'offline' | 'joining' | 'connected';
```
Comment (c43721) [inline src/games/dto/game.dto.ts:15]: nit(s): convert all to enums

File: src/games/gateways/games.gateway.ts [hunk 6]
```diff
@@ -21,12 +32,17 @@ export class GamesGateway implements OnGatewayInit, OnModuleInit {
     private events: Events,
   ) {}
 
+  async emit(event: WebsocketEvent, payload: Serializable<GameDto>) {
+    this.socket.emit(event, await serialize(payload));
+  }
+
   @WsAuthorized()
   @SubscribeMessage('replace player')
+  @UseInterceptors(SerializerInterceptor)
```
Comment (c43721) [inline src/games/gateways/games.gateway.ts:41]: again with global interceptors

File: src/games/gateways/games.gateway.ts [hunk 7]
```diff
@@ -21,12 +32,17 @@ export class GamesGateway implements OnGatewayInit, OnModuleInit {
     private events: Events,
   ) {}
 
+  async emit(event: WebsocketEvent, payload: Serializable<GameDto>) {
+    this.socket.emit(event, await serialize(payload));
+  }
+
   @WsAuthorized()
   @SubscribeMessage('replace player')
```
Comment (c43721) [inline src/games/gateways/games.gateway.ts:40]: a bit of a side-issue but maybe in a follow-up we make these more consistant with FE and make an enum for these

File: src/games/models/game.ts [hunk 8]
```diff
@@ -78,6 +89,41 @@ export class Game extends MongooseDocument {
...
+      map: this.map,
+      state: this.state,
+      connectInfoVersion: this.connectInfoVersion,
+      stvConnectString: this.stvConnectString,
+      logsUrl: this.logsUrl,
+      demoUrl: this.demoUrl,
+      error: this.error,
+      gameServer: this.gameServer
+        ? await gameServersService.getById(this.gameServer)
+        : undefined,
```
Comment (c43721) [inline src/games/models/game.ts:119]: doesn't this return undefined if the id isn't there?

File: src/games/services/__mocks__/games.service.ts [hunk 9]
```diff
@@ -26,7 +26,7 @@ export class GamesService {
   private lastGameId = 0;
 
   constructor(
-    @InjectModel(Game.name) private gameModel: Model<GameDocument>,
+    @InjectModel('Game') private gameModel: Model<GameDocument>,
```
Comment (c43721) [inline src/games/services/__mocks__/games.service.ts:29]: Why this change?

File: src/players/dto/player.dto.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  id: string;
+  name: string;
+  steamId: string;
+  joinedAt: Date;
+  avatar: {
+    small?: string;
+    medium?: string;
+    large?: string;
+  };
+  roles: ('super user' | 'admin' | 'bot')[];
```
Comment (c43721) [inline src/players/dto/player.dto.ts:11]: again with enum- maybe this should be done in a follow-up instead?

File: src/players/gateways/players.gateway.ts [hunk 11]
```diff
@@ -34,6 +38,10 @@ export class PlayersGateway
 
   constructor(private events: Events, private playersService: PlayersService) {}
 
+  async emit(event: WebsocketEvent, payload: Serializable<PlayerDto>) {
```
Comment (c43721) [inline src/players/gateways/players.gateway.ts:41]: we should make this a superclass and inherit this method from that.

should be done in a follow up

**Review by c43721 — APPROVED** [review]
some nits, comments, and other things I found

File: src/auth/strategies/steam.strategy.spec.ts (hunk 1 above)
Comment (garrappachc) [reply src/auth/strategies/steam.strategy.spec.ts]: Hmm, dunno... Good idea, though

File: src/games/controllers/games-with-substitution-requests.controller.ts (hunk 2 above)
Comment (garrappachc) [reply src/games/controllers/games-with-substitution-requests.controller.ts:12]: Good idea

File: src/games/controllers/games-with-substitution-requests.controller.ts (hunk 3 above)
Comment (garrappachc) [reply src/games/controllers/games-with-substitution-requests.controller.ts:13]: I would like this to be explicitly stated in order to catch any involuntary model changes.

File: src/games/gateways/games.gateway.ts (hunk 7 above)
Comment (garrappachc) [reply src/games/gateways/games.gateway.ts:40]: Enum for what? The event names are already an enum

File: src/games/models/game.ts (hunk 8 above)
Comment (garrappachc) [reply src/games/models/game.ts:119]: It will throw an error.

File: src/games/services/__mocks__/games.service.ts (hunk 9 above)
Comment (garrappachc) [reply src/games/services/__mocks__/games.service.ts:29]: Circular imports.

File: src/players/dto/player.dto.ts (hunk 10 above)
Comment (garrappachc) [reply src/players/dto/player.dto.ts:11]: I'd like to avoid using enums in DTO - it's something that should be easy for any API consumers to read.

File: src/games/gateways/games.gateway.ts (hunk 7 above)
Comment (c43721) [reply src/games/gateways/games.gateway.ts:40]: `'replace player'` could be `Events.ReplacePlayer`?

File: src/players/dto/player.dto.ts (hunk 10 above)
Comment (c43721) [reply src/players/dto/player.dto.ts:11]: Not sure if that's a good argument, tbh I think `Role[]` is much more readable.

File: src/games/controllers/games-with-substitution-requests.controller.ts (hunk 3 above)
Comment (c43721) [reply src/games/controllers/games-with-substitution-requests.controller.ts:13]: looks ugly but it makes sense

File: src/games/controllers/games-with-substitution-requests.controller.ts (hunk 2 above)
Comment (c43721) [reply src/games/controllers/games-with-substitution-requests.controller.ts:12]: will PR in follow up I think

File: src/auth/strategies/steam.strategy.spec.ts (hunk 1 above)
Comment (c43721) [reply src/auth/strategies/steam.strategy.spec.ts]: can be done in a follow up

File: src/players/dto/player.dto.ts (hunk 10 above)
Comment (garrappachc) [reply src/players/dto/player.dto.ts:11]: Yeah, but you're forcing the user to lookup the `Role` type, which is not a part of the DTO.

File: src/games/gateways/games.gateway.ts (hunk 7 above)
Comment (garrappachc) [reply src/games/gateways/games.gateway.ts:40]: It's not an event, though - it's websocket call. Nevertheless, it's a good idea.

Comment (c43721) [reply src/games/gateways/games.gateway.ts:40]: Ah. Right. WebSocketEvents
